File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-1.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    unordered_map<Node*, Node*> nodemap;

    Node* copyRandomList(Node* head) {
        // First iteration can't solve the random pointer since we can't reach
        // to the node that haven't been seen.

        // The hashmap can solve for tracing back to node's random

        if (!head) {
            nodemap[nullptr] = nullptr;
            return nullptr;
        }

        Node* copy = new Node(head->val);
        nodemap[head] = copy;
        copy->next = copyRandomList(head->next);
        copy->random = nodemap[head->random];

        return copy;
    }
};

File: Data Structures & Algorithms/clone-graph/submission-1.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        nodemap: Dict[int, Node] = {}

        def dfs(node):
            if node.val in nodemap:
                return nodemap[node.val]

            clone = Node(node.val)
            nodemap[node.val] = clone
            for nb in node.neighbors:
                clone.neighbors.append(dfs(nb))
            
            return clone
        
        if not node:
            return
        return dfs(node)
            
